overlord/snapstate/snapmgr.go

```diff
@@ -365,75 +357,61 @@ func (m *SnapManager) blockedTask(cand *state.Task, running []*state.Task) bool
+func refreshScheduleUsesWeekdays(rs []*timeutil.Schedule) error {
+	for _, s := range rs {
+		if s.Weekday != "" {
+			return fmt.Errorf("%q uses weekdays which is currently not supported", s)
+		}
+	}
+	return nil
 }
 
-// ensureRefreshes ensures that we refresh all installed snaps periodically
-func (m *SnapManager) ensureRefreshes() error {
-	m.state.Lock()
-	defer m.state.Unlock()
-
-	// see if it even makes sense to try to refresh
-	if CanAutoRefresh == nil {
-		return nil
-	}
-	if ok, err := CanAutoRefresh(m.state); err != nil || !ok {
-		return err
-	}
+func (m *SnapManager) getRefreshSchedule() ([]*timeutil.Schedule, error) {
+	refreshScheduleStr := defaultRefreshSchedule
 
 	tr := config.NewTransaction(m.state)
-
-	// allow disabling auto-refresh in the tests
-	if osutil.GetenvBool("SNAPD_DEBUG") {
-		var refreshDisabled bool
-		err := tr.Get("core", "refresh.disabled", &refreshDisabled)
-		if err != nil && !config.IsNoOption(err) {
-			return err
-		}
-		if refreshDisabled {
-			return nil
-		}
-	}
-
-	var lastRefresh time.Time
-	err := tr.Get("core", "refresh.last", &lastRefresh)
+	err := tr.Get("core", "refresh.schedule", &refreshScheduleStr)
 	if err != nil && !config.IsNoOption(err) {
-		return err
+		return nil, err
 	}
-
-	nextRefresh := lastRefresh.Add(minRefreshInterval).Add(m.refreshRandomness)
-	if time.Now().Before(nextRefresh) {
-		return nil
+	refreshSchedule, err := timeutil.ParseSchedule(refreshScheduleStr)
+	if err == nil {
+		err = refreshScheduleUsesWeekdays(refreshSchedule)
 	}
-
-	// check that there is no change in flight already
-	for _, chg := range m.state.Changes() {
-		if chg.Kind() == "auto-refresh" && !chg.Status().Ready() {
-			// change already in motion
-			return nil
+	if err != nil {
+		logger.Noticef("cannot use refresh.schedule: %s", err)
+		refreshSchedule, err = timeutil.ParseSchedule(defaultRefreshSchedule)
+		if err != nil {
+			panic(fmt.Sprintf("defaultRefreshSchedule cannot be parsed: %s", err))
 		}
+		tr.Set("core", "refresh.schedule", defaultRefreshSchedule)
+		tr.Commit()
 	}
 
-	// Check that we have reasonable delays between unsuccessful attempts.
-	// If the store is under stress we need to make sure we do not
-	// hammer it too often
-	if !m.lastRefreshAttempt.IsZero() && m.lastRefreshAttempt.Add(10*time.Minute).After(time.Now()) {
-		return nil
+	// we already have a refresh time, check if we got a new config
+	if !m.nextRefresh.IsZero() {
+		if m.currentRefreshSchedule != refreshScheduleStr {
+			// the refresh schedule has changed
+			logger.Debugf("Option refresh-schedule changed, reloading.")
+			m.nextRefresh = time.Time{}
+		}
 	}
+	m.currentRefreshSchedule = refreshScheduleStr
+
+	return refreshSchedule, nil
+}
 
-	// store attempts in memory so that we can backoff a
+func (m *SnapManager) doAutoRefresh() error {
 	m.lastRefreshAttempt = time.Now()
 	updated, tasksets, err := AutoRefresh(m.state)
 	if err != nil {
+		logger.Noticef("AutoRefresh failed with: %s", err)
```

Comment:
This should probably be a bit more oriented towards the user, since AutoRefresh is an internal method which the user will have no idea about. Perhaps something along the lines of

``"Cannot prepare auto-refresh change: %v"`